src/core/simulation_results.jl

```diff
@@ -1,517 +1,420 @@
-)
-    for name in list
-        date_df = references[stage][name]
-        step_df = DataFrames.DataFrame(
-            Date = Dates.DateTime[],
-            Step = String[],
-            File_Path = String[],
-        )
-        for n in 1:length(step)
-            step_df = vcat(step_df, date_df[date_df.Step .== step[n], :])
-        end
-        results[name] = DataFrames.DataFrame()
-        for (ix, time) in enumerate(step_df.Date)
-            file_path = step_df[ix, :File_Path]
-            var = read_arrow_file("$file_path")
-            results[name] = vcat(results[name], var[1:time_length, :])
+function _validate_names(existing_names::Vector{Symbol}, names::Vector{Symbol})
+    for name in names
+        if name ∉ existing_names
+            @error("$name is not stored", sort!(existing_names))
+            throw(IS.InvalidValue("$name is not stored"))
         end
     end
-    return results
+    nothing
 end
-# internal function to parse through the reference dictionary and grab the file paths
-function _read_references(
-    results::Dict,
-    list::Array,
-    stage::String,
-    references::Dict,
-    time_length::Int,
+
+function _process_timestamps(
+    res::SimulationResults,
+    initial_time::Union{Nothing, Dates.DateTime},
+    count::Union{Int, Nothing},
 )
-    for name in list
-        date_df = references[stage][name]
-        results[name] = DataFrames.DataFrame()
-        for (ix, time) in enumerate(date_df.Date)
-            file_path = date_df[ix, :File_Path]
-            var = read_arrow_file(file_path)
-            var_length = min(time_length, size(var, 1))
-            results[name] = vcat(results[name], var[1:var_length, :])
-        end
+    if initial_time === nothing
+        initial_time = first(get_existing_timestamps(res))
     end
-    return results
+    existing_timestamps = get_existing_timestamps(res)
+    if count === nothing
+        requested_range = [v for v in existing_timestamps if v >= initial_time]
+    else
+        requested_range =
+            collect(range(initial_time, length = count, step = get_interval(res)))
+    end
+    invalid_timestamps = [v for v in requested_range if v ∉ existing_timestamps]
+    if !isempty(invalid_timestamps)
+        throw(IS.InvalidValue(
+            "Timetamps $(invalid_timestamps) not stored",
+            sort!(get_existing_timestamps(res)),
+        ))
+    end
+    return requested_range
 end
-# internal function to remove the overlapping results and only use the most recent
-function _read_time(file_path::String, time_length::Number)
-    time_file_path = joinpath(dirname(file_path), "time_stamp.arrow")
-    temp_time_stamp = read_arrow_file("$time_file_path")
-    time_stamp = temp_time_stamp[(1:time_length), :]
-    time_stamp = convert.(Dates.DateTime, time_stamp)
-    return time_stamp
+
+function _get_variables_values(res::SimulationResults, names::Vector{Symbol}, timestamps)
+    existing_names = get_existing_variables(res)
```

Comment:
For duals and parameters you check for empty names. Because the logic is duplicated in each function it looks like you forgot to do it here. But maybe it was intentional.

Reply:
I forgot and will add it too.